service/lib/agama/dbus/clients/base.rb

```diff
@@ -68,24 +68,37 @@ def service
         # @return [Logger]
         attr_reader :logger
 
-        # Registers callback to be called when the properties of the given object changes
+        # Registers a callback to be called when the properties of the given object changes.
+        #
+        # @param dbus_object [::DBus::Object]
+        # @param block [Proc]
+        def on_properties_change(dbus_object, &block)
+          subscribe(dbus_object, "org.freedesktop.DBus.Properties", "PropertiesChanged", &block)
+        end
+
+        # Subscribes to a D-Bus signal.
         #
         # @note Signal subscription is done only once. Otherwise, the latest subscription overrides
         #   the previous one.
         #
         # @param dbus_object [::DBus::Object]
+        # @param interface [String]
+        # @param signal [String]
         # @param block [Proc]
-        def on_properties_change(dbus_object, &block)
-          @on_properties_change_callbacks ||= {}
-          @on_properties_change_callbacks[dbus_object.path] ||= []
-          @on_properties_change_callbacks[dbus_object.path] << block
+        def subscribe(dbus_object, interface, signal, &block)
+          @signal_handlers ||= {}
+          @signal_handlers[dbus_object.path] ||= {}
+          @signal_handlers[dbus_object.path][interface] ||= {}
+          @signal_handlers[dbus_object.path][interface][signal] ||= []
+          @signal_handlers[dbus_object.path][interface][signal] << block
+
+          callbacks = @signal_handlers.dig(dbus_object.path, interface, signal)
 
-          return if @on_properties_change_callbacks[dbus_object.path].size > 1
+          return if callbacks.size > 1
 
-          dbus_properties = dbus_object["org.freedesktop.DBus.Properties"]
-          dbus_properties.on_signal("PropertiesChanged") do |interface, changes, invalid|
-            callbacks = @on_properties_change_callbacks[dbus_object.path]
-            callbacks.each { |c| c.call(interface, changes, invalid) }
+          interface_proxy = dbus_object[interface]
+          interface_proxy.on_signal(signal) do |iface, changes, invalid|
+            callbacks.each { |c| c.call(iface, changes, invalid) }
```

Comment:
The signature looks like the one for PropertiesChanged (or am I missing something?).

Reply:
Yes, let's simply pass the signal arguments to its handler.